code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,494 @@
+	AddComponent(/datum/component/xenoartifact_pricing)
+	AddComponent(/datum/component/discoverable, 10000, TRUE) //Same values as original artifacts, exploration.
+
+/obj/item/xenoartifact/Initialize(mapload, difficulty)
+	. = ..()
+	material = difficulty //Difficulty is set, in some cases, by xenoartifact_console
+	if(!material)
+		material = pick(BLUESPACE, PLASMA, URANIUM, BANANIUM)
+
+	var/datum/component/xenoartifact_pricing/xenop = GetComponent(/datum/component/xenoartifact_pricing)
+
+	switch(material)
+		if(BLUESPACE)
+			name = "bluespace [name]"
+			generate_traits(list(/datum/xenoartifact_trait/minor/sharp, /datum/xenoartifact_trait/minor/radioactive,
+							/datum/xenoartifact_trait/minor/sentient, /datum/xenoartifact_trait/major/sing, 
+							/datum/xenoartifact_trait/major/laser, /datum/xenoartifact_trait/major/bomb,
+							/datum/xenoartifact_trait/major/handmore, /datum/xenoartifact_trait/major/emp))
+			if(!xenop.price)
+				xenop.price = pick(100, 200, 300)
+
+		if(PLASMA)
+			name = "plasma [name]"
+			generate_traits(list(/datum/xenoartifact_trait/major/sing, /datum/xenoartifact_trait/activator/burn,
+							/datum/xenoartifact_trait/minor/dense, /datum/xenoartifact_trait/minor/sentient, 
+							/datum/xenoartifact_trait/major/capture, /datum/xenoartifact_trait/major/timestop,
+							/datum/xenoartifact_trait/major/bomb, /datum/xenoartifact_trait/major/mirrored,
+							/datum/xenoartifact_trait/major/corginator,/datum/xenoartifact_trait/activator/clock,
+							/datum/xenoartifact_trait/major/invisible,/datum/xenoartifact_trait/major/handmore,
+							/datum/xenoartifact_trait/major/lamp, /datum/xenoartifact_trait/major/forcefield,
+							/datum/xenoartifact_trait/activator/signal,/datum/xenoartifact_trait/major/heal,
+							/datum/xenoartifact_trait/activator/signal, /datum/xenoartifact_trait/activator/batteryneed))
+			if(!xenop.price)
+				xenop.price = pick(200, 300, 500)
+			malfunction_mod = 2
+
+		if(URANIUM)
+			name = "uranium [name]"
+			generate_traits(list(/datum/xenoartifact_trait/major/sing, /datum/xenoartifact_trait/minor/sharp,
+							/datum/xenoartifact_trait/major/laser, /datum/xenoartifact_trait/major/corginator,
+							/datum/xenoartifact_trait/minor/sentient, /datum/xenoartifact_trait/minor/wearable,
+							/datum/xenoartifact_trait/major/handmore, /datum/xenoartifact_trait/major/invisible,
+							/datum/xenoartifact_trait/major/heal), TRUE) 
+			if(!xenop.price)
+				xenop.price = pick(300, 500, 800) 
+			malfunction_mod = 8
+
+		if(BANANIUM)
+			name = "bananium [name]"
+			generate_traits(list(/datum/xenoartifact_trait/major/sing))
+			if(!xenop.price)
+				xenop.price = pick(500, 800, 1000) 
+			malfunction_mod = 0.5
+
+	icon_state = null
+	for(var/datum/xenoartifact_trait/T in traits) //This is kinda weird but it stops certain bugs.
+		if(istype(T, /datum/xenoartifact_trait/minor/dense))
+			T.on_init(src)
+			return
+	for(var/datum/xenoartifact_trait/T in traits)
+		T.on_init(src)
```

Comment:
The comment explains it, it's definitely hacky but, the first loop will 'init' the dense trait and the second will do the rest. This is because doing other traits first and dense last can cause setup issues.